### PR TITLE
refactor!: set LSP8 TokenId Type on deployment / initialization

### PR DESCRIPTION
**Pull request by CJ42**
# What does this PR introduce?

## ⚠️ BREAKING CHANGES

- Add parameter on `constructor` / `initialize` to set the LSP8 Token Id type when the contract is deployed (for standard contracts) or initialized (for proxy contracts).

- Restrict the internal `_setData(bytes32,bytes)` function to disallow to edit the data key `LSP8TokenIdType` after the contract has been deployed or initialized.

## 🚀 Feature

Add Typescript and Solidity constants for the different types of LSP8 Token ID Types

| Value | Type | Description |
| --- | --- | --- |
| 0 | `uint256` | each NFT is represented with a unique number.This number is an incrementing count, where each minted token is assigned the next number.  |
| 1 | `string` | each NFT is represented using a unique name (as a short utf8 encoded string, no more than 32 characters long) |
| 2 | `bytes32` | each NFT is represented using a 32 bytes unique identifier. |
| 3 | `bytes32` | each NFT is represented using a 32 bytes hash digest. |
| 4 | `address` | each NFT is represented as its own ERC725Y smart contract. |

## ♻️ Refactor

Add extra parameter of `uint256 tokenIdType` in `constructor` and `initialize` functions of LSP8 Core, presets and extensions.

Add custom `error` named `LSP8TokenIdType` not editable to disallow setting the data key `LSP8TokenIdType`.

## 🧪 Tests

Add tests to ensure data key / value pair for `LSP8TokenIdType` is set when deploying / initializing a LSP8 contract.

## 📄 Documentation

Update auto-generated Natspec docs for LSP8 contracts.

### PR Checklist

- [x] Wrote Tests
- [x] Wrote & Generated Documentation (readme/natspec/dodoc)
- [x] Ran `npm run lint` && `npm run lint:solidity` (solhint)
- [x] Ran `npm run format` (prettier)
- [x] Ran `npm run build`
- [x] Ran `npm run test`

### CODE REVIEW

**Comment by Hugoo** [comment]
Task linked: [DEV-3725 Set tokenIdType when deploying LSP8](https://app.clickup.com/t/2645698/DEV-3725)

**Review by YamenMerhi — APPROVED** [review]

File: contracts/LSP8IdentifiableDigitalAsset/LSP8IdentifiableDigitalAsset.sol [hunk 1]
```diff
@@ -52,15 +57,27 @@ abstract contract LSP8IdentifiableDigitalAsset is
...
+     * @custom:warning Make sure the tokenId type provided on deployment is correct, as it can only be set once
+     * and cannot be changed in the ERC725Y storage after the contract has been deployed.
      */
     constructor(
         string memory name_,
         string memory symbol_,
-        address newOwner_
-    ) LSP4DigitalAssetMetadata(name_, symbol_, newOwner_) {}
+        address newOwner_,
+        uint256 tokenIdType
```
Comment (b00ste) [inline contracts/LSP8IdentifiableDigitalAsset/LSP8IdentifiableDigitalAsset.sol:74]:
```suggestion
        uint256 tokenIdType_
```

File: contracts/LSP8IdentifiableDigitalAsset/LSP8IdentifiableDigitalAsset.sol [hunk 2]
```diff
@@ -52,15 +57,27 @@ abstract contract LSP8IdentifiableDigitalAsset is
...
+    ) LSP4DigitalAssetMetadata(name_, symbol_, newOwner_) {
+        LSP4DigitalAssetMetadata._setData(
+            _LSP8_TOKENID_TYPE_KEY,
+            abi.encode(tokenIdType)
```
Comment (b00ste) [inline contracts/LSP8IdentifiableDigitalAsset/LSP8IdentifiableDigitalAsset.sol:78]:
```suggestion
            abi.encode(tokenIdType_)
```

File: contracts/LSP8IdentifiableDigitalAsset/LSP8IdentifiableDigitalAsset.sol [hunk 3]
```diff
@@ -52,15 +57,27 @@ abstract contract LSP8IdentifiableDigitalAsset is
 {
     /**
      * @notice Sets the token-Metadata
+     *
+     * @dev Deploy a `LSP8IdentifiableDigitalAsset` contract and set the tokenId type inside the ERC725Y storage of the contract.
+     *
      * @param name_ The name of the token
      * @param symbol_ The symbol of the token
      * @param newOwner_ The owner of the the token-Metadata
```
Comment (b00ste) [inline contracts/LSP8IdentifiableDigitalAsset/LSP8IdentifiableDigitalAsset.sol:65]: Can we add `@param tokenIdType some description`

File: docs/contracts/LSP8IdentifiableDigitalAsset/presets/LSP8CompatibleERC721Mintable.md [hunk 4]
```diff
@@ -31,18 +31,24 @@ When marked as 'public', a method can be called both externally and internally,
...
-| ----------- | :-------: | -------------------------------- |
-| `name_`     | `string`  | The name of the token.           |
-| `symbol_`   | `string`  | The symbol of the token.         |
-| `newOwner_` | `address` | The owner of the token contract. |
+| Name          |   Type    | Description                      |
+| ------------- | :-------: | -------------------------------- |
+| `name_`       | `string`  | The name of the token.           |
+| `symbol_`     | `string`  | The symbol of the token.         |
+| `newOwner_`   | `address` | The owner of the token contract. |
+| `tokenIdType` | `uint256` | -                                |
```
Comment (b00ste) [inline docs/contracts/LSP8IdentifiableDigitalAsset/presets/LSP8CompatibleERC721Mintable.md:51]: missing description

**Review by b00ste — APPROVED** [review]
Everything besides the comments looks good

**Pull request merged**